### PR TITLE
Update TypeCheckAttr to infer what not to differentiate wrt for non-differentiable types.

### PR DESCRIPTION
**Pull request by pschuh**
This happens when inferring default 'wrt:' for @differentiable.

This should resolve: TF-296.

### CODE REVIEW

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@dan-zheng is most familiar with this code path so I'll let him take a look first. My initial speculation is that the original attribute checking code was already checking `Differentiable` conformances, so the logic in this PR could be greatly simplified, but i could be wrong.

**Comment by rxwei** [comment]
> Update TypeCheckAttr to infer @nondiff for non-differentiable types. 

By "infer `@nondiff`" do you just mean "infer what to not differentiate w.r.t."? The `@nondiff` parameter decl attribute (not the type attribute) is deprecated in favor of `wrt:`, and should be removed.

**Comment by pschuh** [comment]
Perhaps? isDifferentiableParam is the only part that overlaps. I suppose I could make a helper lambda that does isDifferentiableParam but takes a bool whether or not to emit errors?

File: lib/Sema/TypeCheckAttr.cpp [hunk 1]
```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
...
+          TC.diagnose(attr->getLocation(),
+                      diag::differentiable_attr_no_currying);
+          return;
+        }
+        for (auto &param : methodTy->getParams()) {
+          allWRTParamTypes.push_back(param.getPlainType());
+        }
       } else {
-        autoDiffParameterIndicesBuilder.setAllParameters();
+        if (auto *fnTy = originalFnTy->getResult()->getAs<AnyFunctionType>()) {
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp]: I wonder if you can reduce duplication with the `if (original->isStatic() || isa<ConstructorDecl>(original))` code path? I think the following may be equivalent:
```c++
auto *methodTy = original->getMethodInterfaceType()->castTo<AnyFunctionType>();
auto *fnTy = originalFnTy->getResult()->getAs<AnyFunctionType>();
```

If you could reduce both branches into one, that would be great!
Maybe sth like:
```cpp
auto *methodTy = originalResultTy->getAs<AnyFunctionType>();
// Dedupe diagnostic.
if (methodTy->getResult()->is<AnyFunctionType>() || !isInstanceMethod || ...) {
  TC.diagnose(attr->getLocation(), diag::differentiable_attr_no_currying);
  return;
}
// Dedupe recording parameter types.
for (auto &param : methodTy->getParams())
  allWrtParamTypes.push_back(param.getPlainType());
// ...
```

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
...
+      // is (Self) -> (Args...) -> R, but we don't want to talk about normal
+      // curried functions.
+      // TODO: Clean all this up.
+      SmallVector<Type, 4> allWRTParamTypes;
       if (original->isStatic() || isa<ConstructorDecl>(original)) {
         auto *methodTy =
             original->getMethodInterfaceType()->castTo<AnyFunctionType>();
-        autoDiffParameterIndicesBuilder
-            .setParameters(0, methodTy->getNumParams());
+        if (methodTy->getResult()->getAs<AnyFunctionType>() != nullptr) {
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp]: Please replace occurrences of `getAs<AnyFunctionTypee>() != nullptr` with `is<AnyFunctionType>()`.
```suggestion
        if (methodTy->getResult()->is<AnyFunctionType>()) {
```

File: lib/Sema/TypeCheckAttr.cpp [hunk 3]
```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
...
+          wrtParamType =
+              whereClauseGenEnv->mapTypeIntoContext(wrtParamInterfaceType);
+        }
+        return hasAssociatedSpaces(wrtParamType);
+      };
+
+      if (original->isStatic() || isa<ConstructorDecl>(original)) {
+        auto *methodTy =
+            original->getMethodInterfaceType()->castTo<AnyFunctionType>();
+        for (unsigned i : range(0, methodTy->getNumParams())) {
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp]: Is this equivalent?
```suggestion
        for (unsigned i : range(methodTy->getNumParams())) {
```

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
...
   // wrt param indices.
   if (!checkedWrtParamIndices) {
     AutoDiffParameterIndicesBuilder autoDiffParameterIndicesBuilder(
         originalFnTy);
     if (parsedWrtParams.empty()) {
+      // This is odd because Self is placed at the end (Args..., Self) due
+      // to existing currying logic. Unfortunately we want to talk about the
+      // underlying method which actually is (Args..., Self), but this function
+      // is (Self) -> (Args...) -> R, but we don't want to talk about normal
+      // curried functions.
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp]: Could you please clarify this comment? The meaning isn't super clear to me.

File: lib/Sema/TypeCheckAttr.cpp [hunk 5]
```diff
@@ -2600,19 +2600,86 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
...
+      // TODO: Clean all this up.
+      SmallVector<Type, 4> allWRTParamTypes;
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp]:
```suggestion
      SmallVector<Type, 4> allWrtParamTypes;
```

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Review by dan-zheng — APPROVED** [review]

**Pull request merged**